> **Tip**: Welcome to the Investigate a Dataset project! You will find tips in quoted sections like this to help organize your approach to your investigation. Before submitting your project, it will be a good idea to go back through your report and remove these sections to make the presentation of your work as tidy as possible. First things first, you might want to double-click this Markdown cell and change the title so that it reflects your dataset and investigation.

# Project: Investigate No-show Appointments in Brazil

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

> **Tip**: In this section of the report, provide a brief introduction to the dataset you've selected for analysis. At the end of this section, describe the questions that you plan on exploring over the course of the report. Try to build your report around the analysis of at least one dependent variable and three independent variables. If you're not sure what questions to ask, then make sure you familiarize yourself with the dataset, its variables and the dataset context for ideas of what to explore.

> If you haven't yet selected and downloaded your data, make sure you do that first before coming back here. In order to work with the data in this workspace, you also need to upload it to the workspace. To do so, click on the jupyter icon in the upper left to be taken back to the workspace directory. There should be an 'Upload' button in the upper right that will let you add your data file(s) to the workspace. You can then click on the .ipynb file name to come back here.

In [19]:
# Use this cell to set up import statements for all of the packages that you
#   plan to use.
# Make sure that visualizations are plotted inline with the notebook
import numpy as np, pandas as pd, matplotlib.pyplot as plt
%matplotlib inline



<a id='wrangling'></a>
## Data Wrangling

> **Tip**: In this section of the report, you will load in the data, check for cleanliness, and then trim and clean your dataset for analysis. Make sure that you document your steps carefully and justify your cleaning decisions.

### General Properties

In [20]:
# Load your data and print out a few lines. 
df = pd.read_csv('no_shows.csv')
df.head()


# Perform operations to inspect data
# types and look for instances of missing or possibly errant data.


PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589978e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262962e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679512e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841186e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

In [21]:
#check for null values
df.isnull().sum()

PatientId         0
AppointmentID     0
Gender            0
ScheduledDay      0
AppointmentDay    0
Age               0
Neighbourhood     0
Scholarship       0
Hipertension      0
Diabetes          0
Alcoholism        0
Handcap           0
SMS_received      0
No-show           0
dtype: int64

#### There are not any null values in any of the columns.

In [22]:
#check for duplicate appointment ids

#it would not make sense if two appointments were given the same id, because the appointment id
#is meant to distinguish each appointment from every other appointment
#if there are two appointments with the same id, this means that the data was entered incorrectly

df['AppointmentID'].duplicated().sum()



0

#### There are not any duplicated appointment ids.

In [23]:
#make sure that none of the binary columns (Scholarship, SMS_received, etc.) have 0 for every row
#if any of these columns has 0 for every row, the column should be deleted from the dataframe
#because the variable in question is not relevant to our analysis

#create a list for all columns with binary entries
binary_col_ls = ['Scholarship', 'Hipertension', 'Diabetes', 'Alcoholism', 'Handcap', 'SMS_received']

#iterate through each binary column to determine the amount of 1 (True) entries in each column

for column in binary_col_ls:
    print(column, (df[column] == 1).sum())
    
#compare this to the number of total entries in each column (total rows in the dataframe)    
print("\n", df.shape[0])



Scholarship 10861
Hipertension 21801
Diabetes 7943
Alcoholism 3360
Handcap 2042
SMS_received 35482

 110527


#### For each binary column, the amount of True (1) entries is greater than zero and less than 110,527.
#### This means that we should retain all of these columns.

In [24]:
#check the data type of the PatientId column by printing the first entry
#patiend id should be an integer
df['PatientId'][0]

29872499824296.0

#### The values in the 'PatientId' column are floats. We will need to change them to integers.

In [25]:
#check the data type of the age column by printing the first entry
#age should be an integer, so that it can be used in computations
df["Age"][0]


62

#### The values in the "Age" column are integers, so there is no need to change the data type.

In [26]:
#check the data type of the ScheduledDay column by printing the first entry
#ScheduledDay should be a datetime
df["ScheduledDay"][0]

'2016-04-29T18:38:08Z'

In [27]:
#check the data type of the AppointmentDay column
#AppointmentDay should also be a datetime
df["AppointmentDay"][0]

'2016-04-29T00:00:00Z'

#### The columns for ScheduledDay and AppointmentDay contain strings instead of datetimes. We will need to fix this.


> **Tip**: You should _not_ perform too many operations in each cell. Create cells freely to explore your data. One option that you can take with this project is to do a lot of explorations in an initial notebook. These don't have to be organized, but make sure you use enough comments to understand the purpose of each code cell. Then, after you're done with your analysis, create a duplicate notebook where you will trim the excess and organize your steps so that you have a flowing, cohesive report.

> **Tip**: Make sure that you keep your reader informed on the steps that you are taking in your investigation. Follow every code cell, or every set of related code cells, with a markdown cell to describe to the reader what was found in the preceding cell(s). Try to make it so that the reader can then understand what they will be seeing in the following cell(s).

### Data Cleaning (Replace this with more specific notes!)

In [ ]:
# After discussing the structure of the data and any problems that need to be
#   cleaned, perform those cleaning steps in the second part of this section.


In [16]:
#Convert the strings in the ScheduledDay column to datetimes
df["ScheduledDay"] = df["ScheduledDay"].apply(pd.to_datetime)

#Verify that the strings have been converted to datetimes
df["ScheduledDay"].head()

0   2016-04-29 18:38:08
1   2016-04-29 16:08:27
2   2016-04-29 16:19:04
3   2016-04-29 17:29:31
4   2016-04-29 16:07:23
Name: ScheduledDay, dtype: datetime64[ns]

#### The ScheduledDay column has been successfully converted into datetimes

In [17]:
#Convert the strings in the AppointmentDay column to datetimes
df["AppointmentDay"] = df["AppointmentDay"].apply(pd.to_datetime)

#Verify that the strings have been converted to datetimes
df["AppointmentDay"].head()

0   2016-04-29
1   2016-04-29
2   2016-04-29
3   2016-04-29
4   2016-04-29
Name: AppointmentDay, dtype: datetime64[ns]

#### AppointmentDay column has been successfully converted into datetimes

<a id='eda'></a>
## Exploratory Data Analysis

> **Tip**: Now that you've trimmed and cleaned your data, you're ready to move on to exploration. Compute statistics and create visualizations with the goal of addressing the research questions that you posed in the Introduction section. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables.

### Research Question 1 (Replace this header name!)

In [ ]:
# Use this, and more code cells, to explore your data. Don't forget to add
#   Markdown cells to document your observations and findings.


### Research Question 2  (Replace this header name!)

In [ ]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed. Make sure that you are clear with regards to the limitations of your exploration. If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work, you should save a copy of the report in HTML or PDF form. Before exporting your report, check over it to make sure that the flow of the report is complete. You should probably remove all of the "Tip" quotes like this one so that the presentation is as tidy as possible. It's also a good idea to look over the project rubric, found on the project submission page at the end of the lesson.

> To export the report to the workspace, you should run the code cell below. If it worked correctly, you should get a return code of 0, and you should see the generated .html file in the workspace directory (click on the jupyter icon in the upper left). Alternatively, you can download the html report via the **File** > **Download as** submenu and then manually upload it to the workspace directory. Once you've done this, you can submit your project by clicking on the "Submit Project" button in the lower right. Congratulations!

In [ ]:
from subprocess import call
call(['python', '-m', 'nbconvert', 'Investigate_a_Dataset.ipynb'])